--- a/libtpm/lib/tpmutil.c
+++ b/libtpm/lib/tpmutil.c
@@ -149,22 +149,6 @@
 			TPM_LowLevel_TransportCharDev_Set();
 #endif
 		break;
-		
-		case TPM_LOWLEVEL_TRANSPORT_TCP_SOCKET:
-			TPM_LowLevel_TransportSocket_Set();
-		break;
-		case TPM_LOWLEVEL_TRANSPORT_UNIXIO:
-#ifdef TPM_POSIX		
-			TPM_LowLevel_TransportUnixIO_Set();
-#endif
-		break;
-		
-
-#ifdef TPM_USE_LIBTPMS
-                case TPM_LOWLEVEL_TRANSPORT_LIBTPMS:
-                        TPM_LowLevel_TransportLibTPMS_Set();
-                break;
-#endif
 	}
 	actual_used_transport = tp;
 
